pkg/resource/managed_reconciler.go

```diff
@@ -466,11 +472,11 @@ func (r *ManagedReconciler) Reconcile(req reconcile.Request) (reconcile.Result,
 		}
 
 		// We've successfully finalized the deletion of our external and managed
-		// resources.
+		// resources. We'll be requeued implicitly (if we still exist) due to
+		// the finalizer deletion and subsequent status update, but otherwise
+		// there's no more work to do.
 		managed.SetConditions(v1alpha1.ReconcileSuccess())
-		// TODO(negz): Why do we requeue here? We've deleted our resource and
-		// removed its finalizer at this stage. There's nothing more to do.
-		return reconcile.Result{RequeueAfter: r.shortWait}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, managed)), errUpdateManagedStatus)
+		return reconcile.Result{Requeue: false}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, managed)), errUpdateManagedStatus)
```

Comment:
Thanks for pointing this out @srueg! It appears to me that this update should not be needed, but is not necessarily going to cause any issues because of the `IgnoreNotFound()` error wrapper.

Reply:
I like @srueg 's suggestion actually. By returning `{Requeue: false}`, we're basically stating that the resource is dead for us. Not much point in calling `Update`, even though `IgnoreNotFound` exists.

Reply:
Also I think there's an issue with etcd and/or the api-server where this weird `StorageError: invalid object, Code: 4` error is returned instead of `NotFound`.

Reply:
@srueg I didn't know that, thanks for the heads-up! Even though these errors do not actually cause much headache in terms of functionality, it's better UX to not have them in the first place if not needed.

Reply:
@muvaf looking back at this again, even though we are returning `{Requeue: false}` if an error is returned we will requeue implicitly. `StorageError` aside for the moment, if somehow we do get an error other than `NotFound`, I don't think we just want to completely ignore it. Even though this should never happen, I think I would rather report an error than saying it shouldn't have had this error so we will pretend that it does not. What do you think?

Reply:
> Also I think there's an issue with etcd and/or the api-server where this weird StorageError: invalid object, Code: 4 error is returned instead of NotFound.

Is this definitely the error returned to our Kubernetes client, or just something we see in the Crossplane logs?

> After some more digging, I think it's ok to call update() in the finalizers, but here it shouldn't be called anymore:

Right. The managed resource reconciler has three responsibilities:

* Orchestrate types that satisfy its interfaces.
* Set the `Synced` status condition.
* Update the managed resource's _status_ (this is distinct from updating the resource metadata and spec).

Finalizers live in metadata, so we given the above context and that that we must update the resource (not its status) I think it's appropriate to call `Update` in the `RemoveFinalizer`.

> but here it shouldn't be called anymore:

I think I agree. The only reason the resource under reconciliation might still exist after `r.managed.Finalize` returns is if some other controller had added additional finalizers to it. When designing this logic I figured that in that edge case we probably wanted to communicate that we processed the delete. I'm happy to remove that logic given that in practice we have no examples of multiple controllers setting a finalizer on a managed resource.

> if somehow we do get an error other than NotFound, I don't think we just want to completely ignore it. Even though this should never happen, I think I would rather report an error than saying it shouldn't have had this error so we will pretend that it does not

I think the idea is that we wouldn't attempt to update the resource status in the happy path where none of our deletion actions return an error, and thus could not get an error when attempting to update the status.